signac/synced_collections/utils.py

```diff
@@ -0,0 +1,219 @@
+    generality, most of these checks use the ABCs defined in :py:mod:`collections.abc`.
+    The price of this flexibility is that `isinstance` checks with these classes
+    are very slow because the ``__instancecheck__`` hooks are implemented in pure
+    Python and require checking many different cases.
+
+    Rather than attempting to directly optimize this behavior, this class provides
+    a workaround by which we can amortize the cost of type checks. Given a set
+    of types that must be resolved and a way to identify each of these (which
+    may be expensive), it maintains a local cache of all instances of a given
+    type that have previously been observed. This reduces the cost of type checking
+    to a simple dict lookup, except for the first time a new type is observed.
+
+    Parameters
+    ----------
+    abstract_type_identifiers : collections.abc.Mapping
+        A mapping from a string identifier for a group of types (e.g. ``"MAPPING"``)
+        to a callable that can be used to identify that type. Due to insertion order
+        guarantees of dictionaries in Python>=3.6 (officially 3.7), it is beneficial
+        to order this dictionary with the most frequently occuring types first.
+
+    Attributes
+    ----------
+    abstract_type_identifiers : Dict[str, Callable[Any, bool]]
+        A mapping from string identifiers for an abstract type to callables that
+        accepts an object and returns True if the object is of the key type and
+        False if not.
+    type_map : Dict[Type, str]
+        A mapping from concrete types to the corresponding named abstract type
+        from :attr:`~.abstract_type_identifiers`.
+
+    """
+
+    def __init__(self, abstract_type_identifiers):
+        self.abstract_type_identifiers = abstract_type_identifiers
+        self.type_map = {}
+
+    def get_type(self, obj):
+        """Get the type string corresponding to this data type.
+
+        Parameters
+        ----------
+        obj : Any
+            Any object whose type to check
+
+        Returns
+        -------
+        str
+            The name of the type, where valid types are the keys of the dict
+            argument to the constructor. If the object's type cannot be identified,
+            will return ``None``.
+
+        """
+        # 0-d NumPy arrays must be handled the right way here.
+        if NUMPY and isinstance(obj, numpy.ndarray):
+            if obj.shape == ():
+                obj = obj.item()
+
+        obj_type = type(obj)
+        enum_type = None
+        try:
+            enum_type = self.type_map[obj_type]
+        except KeyError:
+            for data_type, id_func in self.abstract_type_identifiers.items():
+                if id_func(obj):
+                    enum_type = self.type_map[obj_type] = data_type
+                    break
+            self.type_map[obj_type] = enum_type
+
+        return enum_type
+
+
+def default(o: Any) -> Dict[str, Any]:  # noqa: D102
```

Comment:
1. This function does not _need_ to be public, but it is the API that would be used to encode a SyncedCollection using `orjson` so if we're exposing the `SyncedCollectionJSONEncoder` this would be nice to expose as well.
2. For the same reason (it's the `orjson` name) it would be helping to retain the association. Perhaps `synced_collection_default` to match `SyncedCollectionJSONEncoder`?
3. `default` is the method you have to override, so this cannot be changed.